### PR TITLE
Cover IPv6 healthcheck ranges in --ensure_firewall

### PR DESCRIPTION
**Pull request by sergiitk**
Flag  `--ensure_firewall` creates on-demand (per-test) firewall rule that GCP healthcheck services to request backend health status over IPv4.
This PR adds support for healtchecks over IPv6 ranges by creating another firewall rule.

https://cloud.google.com/load-balancing/docs/firewall-rules

### CODE REVIEW

File: framework/infrastructure/traffic_director.py
Comment (sergiitk) [inline framework/infrastructure/traffic_director.py:1]: - [ ] [grpc/java/v1.62.x/branch/psm-security](https://source.cloud.google.com/results/invocations/ff4bbc39-0772-444d-b0c1-c937acbb406a)
- [ ] [grpc/java/v1.62.x/branch/xds_k8s_lb](https://source.cloud.google.com/results/invocations/4bbdec8b-415d-490b-adac-1a492d12e5e5)
- [ ] [grpc/java/v1.62.x/branch/xds_url_map](https://source.cloud.google.com/results/invocations/8d26247f-b045-4326-a1eb-6504c658020e)

**Comment by sergiitk** [comment]
FYI @larry-safran

File: framework/infrastructure/traffic_director.py [hunk 1]
```diff
@@ -715,32 +718,65 @@ def delete_alternative_forwarding_rule(self, force=False):
...
-            name = self.make_resource_name(self.FIREWALL_RULE_NAME)
-        elif self.firewall_rule:
+        name, name_ipv6 = "", ""
+
+        if self.firewall_rule:
             name = self.firewall_rule.name
-        else:
-            return
+        if self.firewall_rule_ipv6:
+            name_ipv6 = self.firewall_rule.name
```
[the remark this answers is not in the document]
Comment (sergiitk) [reply framework/infrastructure/traffic_director.py:761]: Ah, great catch!

File: framework/infrastructure/traffic_director.py [hunk 2]
```diff
@@ -715,32 +718,65 @@ def delete_alternative_forwarding_rule(self, force=False):
...
+            name_ipv6 = self.make_resource_name(self.FIREWALL_RULE_NAME_IPV6)
+
+        try:
+            if name:
+                self._delete_firewall_rule(name)
+                self.firewall_rule = None
+        finally:
+            if name_ipv6:
+                self._delete_firewall_rule(name_ipv6)
+                self.firewall_rule_ipv6 = None
```
[the remark this answers is not in the document]
Comment (sergiitk) [reply framework/infrastructure/traffic_director.py:775]: Not really. The idea was to attempt to delete ipv6 rule even if ipv4 rule deletion failed.  I'll add a comment explaining what was the point there

[the remark this answers is not in the document]
Comment (sergiitk) [reply framework/infrastructure/traffic_director.py]: Restructured the code, should make more sense now.

File: framework/infrastructure/traffic_director.py
Comment (sergiitk) [reply framework/infrastructure/traffic_director.py:1]: - [ ] [grpc/java/v1.63.x/branch/psm-security](https://source.cloud.google.com/results/invocations/efbbcad6-06ac-41c4-afc5-9b3c1f4c77a6)
- [ ] [grpc/java/v1.63.x/branch/xds_k8s_lb](https://source.cloud.google.com/results/invocations/0d0aeabd-1d8d-4cf8-adf7-d7e9c0159818)
- [ ] [grpc/java/v1.63.x/branch/xds_url_map](https://source.cloud.google.com/results/invocations/9e7db10c-0dcd-481e-8da4-31499e2bff36)

**Review by XuanWang-Amos — APPROVED** [review]

**Pull request merged**